File: buyMaxStockIf_i_stockCanBebought_ith_day.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <map>
#include <queue>
#include<set>
using namespace std;

int maxStock(vector<int> arr, int k) {
    int totalStock = 0;
    int availablePrice = k;
    vector<vector<int>> stockPrice;
    for (int i = 0; i < arr.size(); i++) {
        stockPrice.push_back({ arr[i], i + 1 });
    }
    sort(stockPrice.begin(), stockPrice.end());
    for (int i = 0; i < arr.size(); i++) {
        int price = stockPrice[i][0];
        int stock = stockPrice[i][1];
        if (price * stock <= availablePrice) {
            totalStock += stock;
            availablePrice -= (stock * price);
        }
        else {
            totalStock += (availablePrice/price);
            availablePrice -= price *(availablePrice/price);
        }
    }
    return totalStock;
}
int main() {
    vector<int> price = { 10, 7, 19 };
    cout << maxStock(price, 45);
    
    
	return 0;
}

File: buyMaxStockIf_i_stockCanBebought_ith_day.py
class Pair:
    def __init__(self, price,stock):
        self.price = price
        self.stock = stock
def maxStock(arr, balance):
    totalStock = 0
    v = []
    for i in range(len(arr)):
        v.append(Pair(arr[i], i + 1))
    v = sorted(v, key=lambda p: p.price)
    for i in range(len(arr)):
        if v[i].stock * v[i].price <= balance:
            totalStock += v[i].stock
            balance -= v[i].stock * v[i].price
        else:
            totalStock += balance//v[i].price
            balance -= (balance//v[i].price) * v[i].price
    return totalStock
arr = [10, 7, 19]
print(maxStock(arr, 45))
